### PR TITLE
feat!: Manifest body extractors

### PR DESCRIPTION
**Pull request by piotr-roslaniec**
## What this does

- Fixes: #518 
- Adds Extractors and Predicates functionality for JSON data processing
- Integrates Manifest changes with notary's proof-producing code
- Implements backward compatibility with existing `JsonKey`-based proof creation
- **Contains breaking changes for the Manifest**

## Why it's needed

- Improves flexibility in handling complex JSON structures
- Addresses limitations in the current `JsonKey` approach

## How to test

- Test with various JSON data sources, particularly those that were problematic before
- Verify backward compatibility with existing `JsonKey` implementations
- Test edge cases with nested JSON structures
- Share test examples for additional verification

## Notes for reviewers

Focus areas:
- API ergonomics and intuitiveness
- Code complexity vs functionality balance
- Backward compatibility implementation

Known limitations:
- The current workaround for `JsonKey` compatibility may need future refinement (after a rebase over main)

Added an anonymized Coinbase fixture, courtesy of @drewjenkins 

## Required Reviews

Minimum number of reviews before merge: **2**

### CODE REVIEW

File: web-prover-core/src/http.rs [hunk 1]
```diff
@@ -28,24 +31,19 @@ pub const HTTP_1_1: &str = "HTTP/1.1";
...
+  // TODO: A workaround for backwards-compatibility with JsonKey
+  // TODO: It doesn't afford any security
+  /// Returns the IDs of all extractors in the configuration
+  pub fn json_path(&self) -> Vec<JsonKey> {
+    self
+      .0
+      .extractors
+      .iter()
+      .flat_map(|extractor| extractor.selector.iter().map(|s| JsonKey::String(s.clone())))
+      .collect()
```
Comment (piotr-roslaniec) [inline web-prover-core/src/http.rs:46]: Will be addressed in #543

File: web-prover-core/src/parser/config.rs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+  /// JSON format
+  #[default]
+  Json,
+}
+
+/// The root configuration for data extractors
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
+pub struct ExtractorConfig {
+  /// The format of the data to extract from
+  pub format:     DataFormat,
```
Comment (lonerapier) [inline web-prover-core/src/parser/config.rs:19]: thinking out loud: can we deduce this from the path? like `a.0.b.c` for json and `/html/body/div[2]/article/div[4]/p[21]/b/strong` for xpath?

Comment (piotr-roslaniec) [reply web-prover-core/src/parser/config.rs:19]: I'd prefer to be explicit where possible (but not to the point of redundancy). Nothing is set in stone, perhaps the work on HTML format may influece that #547

**Review by lonerapier — APPROVED** [review]
Very very excited for this!!!

no nits/improvements from me on first pass. Tests are awesome, and self-explanatory.

I'll try to find alternate solutions tomorrow with a more detailed pass (if you do not merge it already by then)

**Comment by lonerapier** [comment]
also can you rebase with base branch, i think it has manifest redundancy changes. will also run CI

**Review by mattes — APPROVED** [review]

**Review by 0xJepsen — APPROVED** [review]

File: web-prover-core/src/parser/extractor.rs [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+  Number,
+  /// Boolean type
+  Boolean,
+  /// Array type
+  Array,
+  /// Object type
+  Object,
+}
+
+impl TryFrom<&str> for ExtractorType {
```
Comment (devloper) [inline web-prover-core/src/parser/extractor.rs:35]: nit: is there some derive macro for this?

**Review by devloper — COMMENTED** [review]
Some nits, nothing major!  The extractor/predicate logic looks solid.  Merge it! 

Let's collab with @drewjenkins to migrate the current integrations to this new format and test that they work.

File: notary/src/tee.rs [hunk 4]
```diff
@@ -252,25 +255,30 @@ pub fn create_tee_proof(
...
-  if !response.matches_client_manifest(&manifest.response) {
-    return Err(NotaryServerError::ManifestResponseMismatch);
-  }
-  Ok(())
+
+  // Check if response matches manifest and extract values
+  response
+    .match_and_extract(&manifest.response)?
+    .filter(|values| !values.is_empty())
+    .ok_or(NotaryServerError::ManifestResponseMismatch)
```
[the remark this answers is not in the document]
Comment (piotr-roslaniec) [reply notary/src/tee.rs:283]: This will be addressed in #511

File: notary/src/tee.rs [hunk 5]
```diff
@@ -252,25 +255,30 @@ pub fn create_tee_proof(
...
 
   Ok(TeeProof { data, signature })
 }
 
-/// Check if `manifest`, `request`, and `response` all fulfill requirements necessary for
-/// a proof to be created
-fn validate_notarization_legal(
+/// Verify that `manifest`, `request`, and `response` fulfill requirements necessary for
+/// a proof to be created and return notarization values if any
+fn get_legal_notarization_values(
```
[the remark this answers is not in the document]
Comment (piotr-roslaniec) [reply notary/src/tee.rs:266]: I wanted to make a distinction between validity and filtering using predicates. Valid (well-formed/expected) values may still not be returned because they fail a predicate.

I will work on communicating this logic and phrasing.
